src/Evolution/DiscontinuousGalerkin/DgElementArray.hpp

```diff
@@ -99,15 +101,17 @@ struct DgElementArray {
       Parallel::CProxy_ConstGlobalCache<Metavariables>& global_cache) noexcept {
     if (next_phase == Metavariables::Phase::RegisterWithObserver) {
       auto& local_cache = *(global_cache.ckLocalBranch());
-      // We currently use a fake temporal id when registering observers but in
-      // the future when we start doing load balancing and elements migrate
+      // We currently use a fake observation_id when registering observers but
```

Comment:
Okay, I don't know what is going on here anymore (and this is not the fault of this PR).  What is being registered with whom? (I think it is the Element components telling the Observer components they will be sending them data, but is it tied with a specific observation? )  Why is it a separate phase?  Why do we need a fake observation_id?  What will happen when Elements are created/destroyed during AMR?  Future load balancing is mentioned to occur at specific times, will the observation_ids no longer be fake then?  If so, what will they represent?  In which case, can't we just make them with the first of what they will represent?  I'm fine converting some/all these questions into an issue in order to get something working

Reply:
* Yes, it is the Element components telling the Observer components they will be sending data.
* Yes, it is tied to a specific observation (the one tagged by `Metavariables::element_observation_type`), which currently means that there can be only one "volume observer" and only one "reduction observer" over the entire set of elements.  A future PR will make the registration mechanism Action-centric rather than Component-centric so that you will be able to have more than one "volume observer" and more than one "reduction observer"; the idea is similar to what I do for registering InterpolationTargets: each Action can have a `register()` function that will register that Action with whomever the Action needs to be registered, and it can have an `observation_type` type alias.
* It is in a separate phase because all registration needs to be completed before any observation can be done, to avoid race conditions. (The Observer needs a count of how many messages it will get, and from whom, so that the Observer knows when it has received enough data to do something).
* We need an observation_id because it encodes what type of observation is being done.  The observation_id also encodes a temporal_id, but that information isn't used (currently) in registration, and this is why it is called "fake".  Before this PR, we passed only a fake temporal_id which was completely ignored.
* For AMR, registration will need to be called again whenever the elements change.
* Similarly, registration will need to be called again at load balancing.
* Last question depends on how AMR and load balancing will work.  If all AMR and load-balancing changes occur at a separate Phase, then registration can be done again with a fake observation_id (in which case we should pass the observation_id's observation_type_hash instead of a fake observation_id and then nothing will be fake).  If AMR and load balancing do not occur at a separate Phase, then the registration mechanism will need to use all of the information in observation_id because there is a possibility of race condition between observing with AMR in configuration A and with AMR in configuration B.

Reply:
- I don't like it being called fake if it does encode some information that is being used.  If #1408 is merged, then can just construct with a signalling_Nan.  Perhaps until that is in `observation_id_with_fake_time`
- I think registration should be done when the Elements are initialized, not in a separate phase.  This way registration/deregistration can be done when Elements are created/destroyed/migrated.  (This can be converted to an issue rather than dealt with in this PR)

Reply:
Re point 2:
I'm not sure I agree (though our disagreement may be in definitions of initialization). I view initialization as setting local data at the beginning of the simulation, which only needs to happen once. As you point out, registration needs to happen several times throughout the evolution. However, I don't think long term it should be a separate phase. We need some way of injecting registration and deregistration into post- and pre-migration. I think we want a separate phase for the first registration, after which all the groups and nodegroups have been set up and we can do registration freely. I think what exactly we can/need to do will be a bit clearer once we have basic migration working.

Reply:
By initialization, I mean the construction of individual Element.  I don't think there should be scattered in different phases (or even different Actions in the same phase), functions that need to be called when an Element is created, otherwise there will be chaos.  I am fine with there being different Actions that are called at the beginning of the simulation or during AMR, but in those single actions, the Element should go from non-existent to completely setup. Otherwise we are going to end up with hard to debug deadlocks when something is forgotten.

Reply:
Yes, I agree. I think it'll be possible to do all registration during the constructor call of an individual AlgorithmImpl for a specific non-group and non-nodegroup component. I think we intentionally do (or can) create all groups and nodegroups first, then singletons and arrays. That would guarantee registration will work. For singletons we may want the ability to create them before, during, or after the group phase, but until we have a use for that level of flexibility I say we hold off on it. Would that implementation address your concerns?